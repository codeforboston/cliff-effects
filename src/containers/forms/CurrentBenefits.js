import { connect } from 'react-redux';

import { CurrentBenefitsStep } from '../../forms/CurrentBenefits';
import { setHasBenefit } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),

    USState: state.getIn([
      'geography',
      'state',
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setHasBenefit({ benefit, value }) {
      dispatch(
        setHasBenefit({
          time: 'current',
          benefit,
          value,
        })
      );
    },
  };
};

const CurrentBenefitsContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentBenefitsStep);

export { CurrentBenefitsContainer as CurrentBenefitsStep };
