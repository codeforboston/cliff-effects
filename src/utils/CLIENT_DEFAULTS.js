import { cloneDeep } from 'lodash';

/**
 * Default client values.
 * MUST ALWAYS BE VALID
 * 
 * Note: A 'positive number' money value can have more than
 *     two decimal places.
 */
const CLIENT_DEFAULTS = {

  // Property Name                | Default Value   | Valid Value Description

  current: {
    // Current programs
    hasSnap:                        false,            // boolean
    hasHousing:                     false,            // boolean
    // Household
    // 'm_' for 'member'
    household:  [{                  // 'head' member  // one or more member objects
        m_age:                      30,               // whole number
        m_role:                     'head',           // 'head', 'spouse', 'member'
        m_disabled:                 false             // boolean
    }],
    // Income
    earned:                         0,                // positive number (can have more than two decimal places)
    TAFDC:                          0,                // positive number
    SSI:                            0,                // positive number
    SSDI:                           0,                // positive number
    childSupportIn:                 0,                // positive number
    unemployment:                   0,                // positive number
    workersComp:                    0,                // positive number
    pension:                        0,                // positive number
    socialSecurity:                 0,                // positive number
    alimony:                        0,                // positive number
    otherIncome:                    0,                // positive number
    incomeExclusions:               0,                // positive number
    // Expenses
    childDirectCare:                0,                // positive number
    childBeforeAndAfterSchoolCare:  0,                // positive number
    childTransportation:            0,                // positive number
    childOtherCare:                 0,                // positive number
    earnedBecauseOfChildCare:       0,                // positive number
    childSupportPaidOut:            0,                // positive number
    adultDirectCare:                0,                // positive number
    adultTransportation:            0,                // positive number
    adultOtherCare:                 0,                // positive number
    disabledAssistance:             0,                // positive number
    earnedBecauseOfAdultCare:       0,                // positive number
    disabledMedical:                0,                // positive number
    otherMedical:                   0,                // positive number
    /** @todo When client has section 8, switch this to 'housingVoucher' */
    /** @todo Make shelter expense values more robust. */
    shelter:                        'homeless',       // 'homeless', 'homeowner', 'renter', 'housingVoucher'
    contractRent:                   0,                // positive number
    rentShare:                      0,                // positive number
    rent:                           0,                // positive number
    mortgage:                       0,                // positive number
    housingInsurance:               0,                // positive number
    propertyTax:                    0,                // positive number
    climateControl:                 false,            // boolean
    nonHeatElectricity:             false,            // boolean
    phone:                          false,            // boolean
    fuelAssistance:                 false,            // boolean  Note: Input is 'yes', 'no', but should be being converted
    otherExpenses:                  0                 // positive number
  },

  // Note: A `future` value will default to the client's `current` value until a user changes it explicitly
  future: {},                       // same structure // same

};

CLIENT_DEFAULTS.future = cloneDeep( CLIENT_DEFAULTS.current );


export { CLIENT_DEFAULTS };
