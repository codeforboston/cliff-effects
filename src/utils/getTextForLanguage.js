/** Returns a translator based on the language given */

// TRANSFORMER FUNCTIONS
import { mergeWith } from 'lodash';
import { interpolateSnippets } from './interpolation.js';

// DATA
import { getLocalizationData } from '../localization/all';
import inlineComponents from '../localization/inlineComponents';

// Get copy of localization data
const localizations = getLocalizationData();

// store interpolated and (if necessary) merged snippets objects
let finishedSnippets = { en: interpolateSnippets(localizations.en, inlineComponents) };

/** Customizes Lodash's mergeWith function to replace arrays completely
 * (to avoid arrays of English strings being mixed with arrays of translated
 * strings, if they happen to have different lengths).
 */
const mergeCustomizer = function (objValue, srcValue) {
  if (Array.isArray(objValue)) {
    return srcValue;
  }
};

/** Returns the object named by langName that contains
 * the text snippets of that language. If that language
 * doesn't exist, it warns the coder and returns English.
 */
const getTextForLanguage = function (langCode) {
  if (!localizations[ langCode ]) {
    console.warn('There\'s no localization for ' + langCode + '. Defaulting to English.');
    langCode = 'en';
  }

  if (!finishedSnippets[ langCode ]) {
    // interpolate snippets and merge with English (filling in any gaps)
    const interpolatedSnippets = interpolateSnippets(localizations[ langCode ], inlineComponents);

    // deeply merge the object containing snippets in langName with English,
    // so that we fall back to English if a particular field is missing.
    finishedSnippets[ langCode ] =  mergeWith({}, finishedSnippets.en, interpolatedSnippets, mergeCustomizer);
  }

  return finishedSnippets[ langCode ];
};  // End getTextForLanguage()


export { getTextForLanguage };
