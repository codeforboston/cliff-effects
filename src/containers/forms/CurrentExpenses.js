import { connect } from 'react-redux';

import { CurrentExpensesStep } from '../../forms/CurrentExpenses';

import { setCashValue, setHousingType, setPaysUtility, setGetsFuelAssistance } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setExpenseValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },

    setHousingType({ housingType }) {
      dispatch(
        setHousingType({
          time: 'current',
          housingType,
        })
      );
    },

    setPaysUtility({ utility, paysUtility }) {
      dispatch(
        setPaysUtility({
          time: 'current',
          utility,
          paysUtility,
        })
      );
    },

    setGetsFuelAssistance({ getsAssistance }) {
      dispatch(
        setGetsFuelAssistance({
          time: 'current',
          getsAssistance,
        })
      );
    },
  };
};

const CurrentExpensesStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentExpensesStep);

export { CurrentExpensesStepContainer as CurrentExpensesStep };
