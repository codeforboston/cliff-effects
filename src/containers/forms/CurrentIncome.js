import { connect } from 'react-redux';

import { CurrentIncomeStep } from '../../forms/CurrentIncome';

import { setCashValue } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setIncomeValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },
  };
};

const CurrentIncomeStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentIncomeStep);

export { CurrentIncomeStepContainer as CurrentIncomeStep };
