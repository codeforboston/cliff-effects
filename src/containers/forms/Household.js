import { connect } from 'react-redux';

import { HouseholdStep } from '../../forms/Household';
import {
  addMember,
  removeMember,
  setMemberIsDisabled,
  setMemberRole,
  setMemberAge,
} from '../../actions';

const time = 'current';

const mapStateToProps = (state) => {
  return {
    household: state.getIn([
      'client',
      'current',
      'household', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setMemberAge({ index, age }) {
      dispatch(
        setMemberAge({
          time,
          index,
          age,
        })
      );
    },

    setMemberIsDisabled({ index, isDisabled }) {
      dispatch(
        setMemberIsDisabled({
          time,
          index,
          isDisabled,
        })
      );
    },

    setMemberRole({ index, role }) {
      dispatch(
        setMemberRole({
          time,
          index,
          role,
        })
      );
    },

    removeMember({ index }) {
      dispatch(
        removeMember({
          time,
          index,
        })
      );
    },

    addMember({ member }) {
      dispatch(
        addMember({
          time,
          member,
        })
      );
    },
  };
};

const HouseholdStepContainer = connect(mapStateToProps, mapDispatchToProps)(HouseholdStep);

export { HouseholdStepContainer as HouseholdStep };
