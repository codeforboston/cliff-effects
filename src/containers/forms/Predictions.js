import { connect } from 'react-redux';

import { PredictionsStep } from '../../forms/Predictions';

import { setCashValue } from '../../actions';

const mapStateToProps = (state) => {
  return { client: state.get('client') };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setPredictionValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'future',
          name,
          value,
        })
      );
    },
  };
};

const PredictionsStepContainer = connect(mapStateToProps, mapDispatchToProps)(PredictionsStep);

export { PredictionsStepContainer as PredictionsStep };
