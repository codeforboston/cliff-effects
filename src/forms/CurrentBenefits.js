// REACT COMPONENTS
import React from 'react';
import { Form } from 'semantic-ui-react';

// PROJECT COMPONENTS
import { FormPartsContainer, ControlledRadioYesNo } from './formHelpers';

// COMPONENT HELPER FUNCTIONS
import { getTimeSetter } from '../utils/getTimeSetter';


const LocalizedRadioYesNo = function ({ snippets, checked, name, onChange }) {

  return (
    <ControlledRadioYesNo
      checked   = { checked }
      labelText = { snippets[ name ][ 'label' ] }
      name      = { name }
      onChange  = { onChange } />
  );
};


/**
 * @todo Add 'vertical list of options' creator that will create a list of fields using the `.field-aligner` class
 *
 * @function
 * @param {object} props See below.
 * @property {object} props.current Client current info.
 * @property {function} props.setClientProperty Updates state upstream.
 * @property {function} props.snippets Uses user chosen language-specific
 *    snippets.
 *
 * @returns {object} Component
 */
const CurrentBenefitsContent = ({ current, setClientProperty, snippets }) => {

  return (
    <div >
      <LocalizedRadioYesNo
        checked   = { current.hasSection8 }
        name      = { 'hasSection8' }
        onChange  = { setClientProperty }
        snippets  = { snippets } />
      <LocalizedRadioYesNo
        checked   = { current.hasSnap }
        name      = { 'hasSnap' }
        onChange  = { setClientProperty }
        snippets  = { snippets } />
    </div>
  );  // end return

};  // End CurrentBenefitsContent()

/** @todo Abstract all the step components?
 *
 * @function
 * @param {object} props See below.
 * @property {function} props.changeClient Updates state upstream.
 * @property {function} props.snippets Uses user chosen language-specific
 *    snippets.
 * @property {object} props.client JSON object with future and current values.
 * @property {function} props.nextStep Go to next form section.
 *
 * @returns {object} Component
 */
const CurrentBenefitsStep = ({ changeClient, nextStep, client, snippets }) => {

  /** @todo Abstract `getTimeSetter()` use to VisitPage.js? */
  const setTimeProp = getTimeSetter('current', changeClient);

  return (
    <Form
      size='massive'
      className='household-size-form flex-item flex-column'>
      <FormPartsContainer
        title     = { snippets.currentBenefits }
        clarifier = { snippets.selectBenefits }
        right     = {{ name: 'Next', func: nextStep }}>
        <CurrentBenefitsContent
          setClientProperty = { setTimeProp }
          current           = { client.current }
          snippets          = { snippets } />
      </FormPartsContainer>

    </Form>
  );

};  // End CurrentBenefitsStep()

export { CurrentBenefitsStep };
