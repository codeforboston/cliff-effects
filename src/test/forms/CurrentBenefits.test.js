import React from 'react';
import { shallow } from 'enzyme';

import { CurrentBenefitsStep } from '../../forms/CurrentBenefits';

import { CLIENT_DEFAULTS } from '../../utils/CLIENT_DEFAULTS';

// LOCALIZATION
import { getTextForLanguage } from '../../utils/getTextForLanguage';


var snippets = getTextForLanguage(`en`);
test('Benefits step component renders as snapshot correctly', () => {
  const navData = {
          left:   null,
          middle: null,
          right:  { text: 'Next', onClick: jest.fn() },
        },
        updateClientValue     = jest.fn(),
        saveForm         = jest.fn(),
        askToResetClient = jest.fn(),
        openFeedback     = jest.fn(),
        formSnippets     = snippets.visitPage.currentBenefits;

  const wrapper = shallow(
    <CurrentBenefitsStep
      currentStep       = { 1 }
      client            = { CLIENT_DEFAULTS }
      navData           = { navData }
      updateClientValue      = { updateClientValue }
      saveForm          = { saveForm }
      askToResetClient  = { askToResetClient }
      openFeedback      = { openFeedback }
      snippets          = { formSnippets } />
  );

  expect(wrapper).toMatchSnapshot();
});
