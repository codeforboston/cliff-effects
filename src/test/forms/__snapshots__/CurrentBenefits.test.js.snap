// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Benefits step component renders as snapshot correctly 1`] = `
<FormPartsContainer
  clarifier={
    <span
      className="snippet"
      lang="en"
    >
      Select the benefits you currently receive
    </span>
  }
  formClass="benefits"
  formSize="massive"
  navData={
    Object {
      "left": null,
      "middle": null,
      "right": Object {
        "onClick": [Function],
        "text": "Next",
      },
    }
  }
  title={
    <span
      className="snippet"
      lang="en"
    >
      Current Benefits
    </span>
  }
>
  <CurrentBenefitsContent
    current={
      Object {
        "SSDI": 0,
        "SSI": 0,
        "TAFDC": 0,
        "adultDirectCare": 0,
        "adultOtherCare": 0,
        "adultTransportation": 0,
        "alimony": 0,
        "childBeforeAndAfterSchoolCare": 0,
        "childDirectCare": 0,
        "childOtherCare": 0,
        "childSupportIn": 0,
        "childSupportPaidOut": 0,
        "childTransportation": 0,
        "climateControl": false,
        "contractRent": 0,
        "disabledAssistance": 0,
        "disabledMedical": 0,
        "earned": 0,
        "earnedBecauseOfAdultCare": 0,
        "earnedBecauseOfChildCare": 0,
        "fuelAssistance": false,
        "hasSection8": false,
        "hasSnap": false,
        "household": Array [
          Object {
            "m_age": 30,
            "m_disabled": false,
            "m_role": "head",
          },
        ],
        "housing": "homeless",
        "housingInsurance": 0,
        "mortgage": 0,
        "nonHeatElectricity": false,
        "otherExpensesCable": 0,
        "otherExpensesCareProducts": 0,
        "otherExpensesClothes": 0,
        "otherExpensesEntertainment": 0,
        "otherExpensesFood": 0,
        "otherExpensesMedical": 0,
        "otherExpensesOther": 0,
        "otherExpensesPhone": 0,
        "otherExpensesTransport": 0,
        "otherExpensesUtilities": 0,
        "otherIncome": 0,
        "otherMedical": 0,
        "pension": 0,
        "phone": false,
        "propertyTax": 0,
        "rent": 0,
        "rentShare": 0,
        "socialSecurity": 0,
        "unemployment": 0,
        "wantsToSeeOtherExpenses": false,
        "workersComp": 0,
      }
    }
    snippets={
      Object {
        "i_currentBenefits": <span
          className="snippet"
          lang="en"
      >
          Current Benefits
      </span>,
        "i_hasSection8Hint": <span
          className="snippet"
          lang="en"
      >
          Section 8 provides rental housing assistance.
      </span>,
        "i_hasSection8Label": <span
          className="snippet"
          lang="en"
      >
          Do you have Section 8 Housing?
      </span>,
        "i_hasSnapHint": <span
          className="snippet"
          lang="en"
      >
          SNAP provides assistance with buying food
      </span>,
        "i_hasSnapLabel": <span
          className="snippet"
          lang="en"
      >
          Do you have SNAP?
      </span>,
        "i_selectBenefits": <span
          className="snippet"
          lang="en"
      >
          Select the benefits you currently receive
      </span>,
        "langCode": "en",
      }
    }
    updateClientValue={[Function]}
  />
</FormPartsContainer>
`;
